internal/lib/sandbox/namespaces_linux.go

```diff
@@ -105,22 +125,57 @@ func pinNamespaces(nsTypes []NSType, cfg *config.Config) ([]NamespaceIface, erro
 		return nil, fmt.Errorf("failed to pin namespaces %v: %s %v", nsTypes, output, err)
 	}
 
-	returnedNamespaces := make([]NamespaceIface, 0)
+	returnedNamespaces := make([]NamespaceIface, 0, numNSToPin)
 	for _, info := range mountedNamespaces {
-		ret, err := nspkg.GetNS(info.path)
+		iface, err := info.toIface()
 		if err != nil {
 			return nil, err
 		}
-
-		returnedNamespaces = append(returnedNamespaces, &Namespace{
-			ns:     ret.(NS),
-			nsType: info.nsType,
-			nsPath: info.path,
-		})
+		returnedNamespaces = append(returnedNamespaces, iface)
 	}
 	return returnedNamespaces, nil
 }
 
+func pinPidNamespace(cfg *config.Config, path string) (iface NamespaceIface, errRet error) {
+	// verify the path we were passed is indeed a namespace
+	if err := nspkg.IsNSorErr(path); err != nil {
+		return nil, err
+	}
+	pinnedNamespace := uuid.New().String()
+	namespaceInfo := newNamespaceInfo(cfg.NamespacesDir, pinnedNamespace, PIDNS)
+
+	// ensure the parent directory is there
+	if err := os.MkdirAll(filepath.Join(cfg.NamespacesDir, "pidns"), 0o755); err != nil {
+		return nil, err
```

Comment:
Let’s wrap those errors too. The amount of information the original errors provide is low.